# Python for Scientific Research
# Flow control
# Answers to exercises

## Exercise 1

Write a script that finds out whether a certain year (e.g., `2004`, `1600` or `51`; contained in a variable `year`) is a leap year. A year is said to be leap year if the year is exactly divisible by 4 but not divisible by 100. Year is also a leap year if it is exactly divisible by 400.

If the year is a leap year, the script should print "The year XXXX is a leap year", otherwise it should print "Nope, the year XXXX is not a leap year".

You can assume that the value contained in `year` is a positive integer.

The [table on operator precedence](https://docs.python.org/3/reference/expressions.html?highlight=precedence#operator-precedence) may be useful here again.

In [1]:
year = 1900

# note that operator precedence 
if year % 4 == 0 and year % 100 != 0 or year % 400 == 0:
    print(f"The year {year} is a leap year.")
else:
    print(f"Nope, the year {year} is not a leap year.")

Nope, the year 1900 is not a leap year.


## Exercise 2

### 2.1 Loop from 1 to 10 and print out the square of that number 

In [2]:
for i in range(1,11):
    print(i**2)

1
4
9
16
25
36
49
64
81
100


### 2.2 GC content of random DNA sequences

Create a list of 5 short DNA sequences (e.g `["acccgtagcgta", "cgaattgccgaa", ...]`). Loop through this list and for each DNA sequence compute the total numbers of "g" and "c" basepairs, and subsequently the percentage GC content (i.e., the percentage of basepairs that are guanine or cytosine). Print the information out. 

In [3]:
sequences = ["acccgtagcgta","cgaattgccgaa","ggggggg","cccccccc","aaaaaaatttttt"]

for seq_i in sequences:
    
    # get total number of Gs
    count_g = seq_i.count("g")
    
    # get total number of Cs
    count_c = seq_i.count("c") 
    
    # calculate total length of the string
    total_bp = len(seq_i) 
    
    # calculate GC percentage
    gc_percentage = (count_g + count_c)/total_bp 
    
    # print out the information
    print(f"Sequence {seq_i} contains {count_g} Gs, {count_c} Cs and its GC content is {gc_percentage}")

Sequence acccgtagcgta contains 3 Gs, 4 Cs and its GC content is 0.5833333333333334
Sequence cgaattgccgaa contains 3 Gs, 3 Cs and its GC content is 0.5
Sequence ggggggg contains 7 Gs, 0 Cs and its GC content is 1.0
Sequence cccccccc contains 0 Gs, 8 Cs and its GC content is 1.0
Sequence aaaaaaatttttt contains 0 Gs, 0 Cs and its GC content is 0.0


## Exercise 3
### Exercise 3.1
Using list comprehensions, populate a list with the square of numbers from 1 to 10 (i.e 1, 4, 9, etc.)

In [4]:
squares = [i**2 for i in range(1,11)]
print(squares)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


### Exercise 3.2
Using list comprehensions, populate a list, elements of which are tuples (i,j,k) where i is a number from 1 to 10 and j is its square,while k is its square root

In [5]:
numberCollection = [(i,i**2,i**0.5) for i in range(1,11)]
print(numberCollection)

[(1, 1, 1.0), (2, 4, 1.4142135623730951), (3, 9, 1.7320508075688772), (4, 16, 2.0), (5, 25, 2.23606797749979), (6, 36, 2.449489742783178), (7, 49, 2.6457513110645907), (8, 64, 2.8284271247461903), (9, 81, 3.0), (10, 100, 3.1622776601683795)]


### Exercise 3.3
Subsequently, use a separate loop to print out the contents of each tuple as follows:  
"The square of 1 is 1 while its square root is 1.0"  
"The square of 2 is 4 while its square root is 1.414214"  
etcetera  

(Hint: to get the proper precision of the square root, use the `g` [formatting modifier](https://docs.python.org/3/library/string.html?highlight=str%20format#format-specification-mini-language))

In [6]:
for numTuple in numberCollection:
    print(f"The square of {numTuple[0]} is {numTuple[1]} while its square root is {numTuple[2]:g}")

The square of 1 is 1 while its square root is 1
The square of 2 is 4 while its square root is 1.41421
The square of 3 is 9 while its square root is 1.73205
The square of 4 is 16 while its square root is 2
The square of 5 is 25 while its square root is 2.23607
The square of 6 is 36 while its square root is 2.44949
The square of 7 is 49 while its square root is 2.64575
The square of 8 is 64 while its square root is 2.82843
The square of 9 is 81 while its square root is 3
The square of 10 is 100 while its square root is 3.16228
